File: legacy/react-mvp/src/lib/csv-export.ts
import {
  getAction,
  getCategory,
  getPlayer,
  getTeam,
  listEventsByMatch,
} from '../db/repository';
import type { Event } from '../types';

export interface CsvRow {
  period: number;
  time: number;
  subject: string;
  category: string;
  action: string;
  outcome: string;
  comment: string;
}

const CSV_HEADERS = [
  'Тайм',
  'Время',
  'Игрок/Команда',
  'Категория',
  'Действие',
  'Результат',
  'Комментарий',
] as const;

function escapeCsv(value: string): string {
  if (value.includes(',') || value.includes('"') || value.includes('\n')) {
    return `"${value.replace(/"/g, '""')}"`;
  }
  return value;
}

export async function buildCsvRows(matchId: number): Promise<CsvRow[]> {
  const events = await listEventsByMatch(matchId);
  const rows: CsvRow[] = [];

  for (const event of events) {
    const action = event.actionId ? await getAction(event.actionId) : undefined;
    const category = action?.categoryId ? await getCategory(action.categoryId) : undefined;

    let subject = '';
    if (event.subjectType === 'player' && event.playerId) {
      const player = await getPlayer(event.playerId);
      subject = player?.name ?? '';
    } else if (event.subjectType === 'team' && event.teamId) {
      const team = await getTeam(event.teamId);
      subject = team?.name ?? '';
    }

    rows.push({
      period: event.periodNumber,
      time: event.timestampSec,
      subject,
      category: category?.name ?? '',
      action: action?.name ?? '',
      outcome: event.outcome ?? '',
      comment: event.comment ?? '',
    });
  }

  return rows;
}

export function rowsToCsv(rows: CsvRow[]): string {
  const lines = [CSV_HEADERS.join(',')];
  for (const row of rows) {
    lines.push(
      [
        String(row.period),
        String(row.time),
        escapeCsv(row.subject),
        escapeCsv(row.category),
        escapeCsv(row.action),
        escapeCsv(row.outcome),
        escapeCsv(row.comment),
      ].join(','),
    );
  }
  return lines.join('\n');
}

export async function exportMatchToCsv(matchId: number, filename: string): Promise<void> {
  const rows = await buildCsvRows(matchId);
  const csv = rowsToCsv(rows);
  const blob = new Blob(['\uFEFF' + csv], { type: 'text/csv;charset=utf-8;' });
  const link = document.createElement('a');
  link.href = URL.createObjectURL(blob);
  link.download = filename;
  document.body.appendChild(link);
  link.click();
  document.body.removeChild(link);
  URL.revokeObjectURL(link.href);
}

export async function getEventsForMatch(matchId: number): Promise<Event[]> {
  return listEventsByMatch(matchId);
}

File: backend/csv_export.py
from __future__ import annotations

import csv
import io
import re
import sqlite3
from urllib.parse import quote

from . import repository as repo

CSV_HEADERS = ["Тайм", "Время", "Игрок/Команда", "Категория", "Действие", "Результат", "Комментарий"]


def _sanitize_filename(name: str) -> str:
    sanitized = re.sub(r'[<>:"/\\|?*\x00-\x1f]+', "_", name).strip().strip(".")
    return sanitized or "match_export"


def build_export_filename(conn: sqlite3.Connection, match_id: int) -> str:
    match = repo.get_match(conn, match_id)
    if not match:
        return "match_export.csv"

    home = repo.get_team(conn, int(match["HomeTeamId"]))
    away = repo.get_team(conn, int(match["AwayTeamId"]))
    home_name = str(home["Name"]) if home else "home"
    away_name = str(away["Name"]) if away else "away"
    tournament = (match["TournamentName"] or "").strip()

    parts = [f"{home_name} vs {away_name}"]
    if tournament:
        parts.append(tournament)
    return _sanitize_filename(" - ".join(parts)) + ".csv"


def export_content_disposition(filename: str) -> str:
    try:
        filename.encode("ascii")
        return f'attachment; filename="{filename}"'
    except UnicodeEncodeError:
        fallback = "match_export.csv"
        return f"attachment; filename=\"{fallback}\"; filename*=UTF-8''{quote(filename)}"


def build_csv(conn: sqlite3.Connection, match_id: int) -> str:
    events = repo.list_events_by_match(conn, match_id)
    output = io.StringIO()
    writer = csv.writer(output)
    writer.writerow(CSV_HEADERS)

    for event in events:
        action = repo.get_action(conn, int(event["ActionId"])) if event["ActionId"] else None
        category = repo.get_category(conn, int(action["CategoryId"])) if action else None
        subject = ""
        if event["SubjectType"] == "player" and event["PlayerId"]:
            player = repo.get_player(conn, int(event["PlayerId"]))
            subject = player["Name"] if player else ""
        elif event["SubjectType"] == "team" and event["TeamId"]:
            team = repo.get_team(conn, int(event["TeamId"]))
            subject = team["Name"] if team else ""

        writer.writerow(
            [
                event["PeriodNumber"],
                event["TimestampSec"],
                subject,
                category["Name"] if category else "",
                action["Name"] if action else "",
                event["Outcome"] or "",
                event["Comment"] or "",
            ]
        )

    return "\ufeff" + output.getvalue()

